Add driver for LT7170/LT7170-1, LT7171/LT7171-1

csarip opened the pull request:
## Pull Request Description

The LT7170/LT7170-1 are monolithic DC/DC synchronous step-down regulators that 
deliver up to 20 A of continuous output current. The LT7170-1 option has two 
switching phases that are connected to two inductors to drive a 
single-regulated output supply.

Similarly, the LT7171/LT7171-1 are monolithic PolyPhase DC/DC synchronous 
step-down regulators that also deliver up to 20A of continuous output current. 
The LT7171-1 option has two switching phases that are connected to two 
inductors to drive a single-regulated output supply.

Both series offer quick, clean, low-overshoot switching edges deliver high 
efficiency while minimizing electromagnetic interference (EMI) emissions.

An I2C-based PMBus 1.3-compliant serial interface allows control of device 
functions while providing telemetry information for system monitoring. The 
LT7170/LT7170-1 and LT7171/LT7171-1 are supported by the LTpowerPlay graphical 
user interface (GUI) tool.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2299) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2299) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=2299) it.</sub>

kister-jimenez commented:
/AzurePipelines run

kister-jimenez commented:
/AzurePipelines run

kister-jimenez commented:
/AzurePipelines run

kister-jimenez commented on drivers/power/lt7170/lt7170.c:
```diff
@@ -0,0 +1,1074 @@
+	*reg = no_os_field_prep(LT7170_IEEE754_SIGN_BIT, sign) |
+	       no_os_field_prep(LT7170_IEEE754_MANTISSA_MSK, mantissa) |
+	       no_os_field_prep(LT7170_IEEE754_EXPONENT_MSK, exponent);
+
+	return 0;
+}
+/**
+ * @brief Converts raw IEEE754 register data to its actual value
```
nit: add a newline between functions.

kister-jimenez commented on drivers/power/lt7170/lt7170.c:
```diff
@@ -0,0 +1,1074 @@
+		goto free_block;
+	}
+
+	/* Initialize I2C */
```
You can probably remove this comment as this is obvious.

kister-jimenez commented on drivers/power/lt7170/lt7170.c:
```diff
@@ -0,0 +1,1074 @@
+	ret = lt7170_read_block_data(dev, LT7170_IC_DEVICE_ID, block,
+				     lt7170_info[init_param->chip_id].name_size);
+	if (ret)
+		goto dev_err;
```
nit: adding a newline after the goto line makes this easier to read.

kister-jimenez commented on drivers/power/lt7170/lt7170.c:
```diff
@@ -0,0 +1,1074 @@
+ * 		 4000 kHz.
+ * @return 0 in case of success, negative error code otherwise
+ */
+int lt7170_switch_freq(struct lt7170_dev *dev, int freq) // freq in kHz
```
No need for the inline comment. freq is describe in the @param.

kister-jimenez commented on drivers/power/lt7170/lt7170.c:
```diff
@@ -0,0 +1,1074 @@
+/**
+ * @brief Set switching frequency
+ * @param dev - Device structure
+ * @param freq - Frequency to set in Hz. Value should be between 400 kHz and
+ * 		 4000 kHz.
```
Confusing parameter. Is it Hz or kHz?

csarip replied:
It is set in Hz. Would it be better if I avoid using 'k' and display the zeros here instead?

csarip replied:
It is set in Hz. Would it be better if I avoid using 'k' and display the zeros here instead?

kister-jimenez commented on drivers/power/lt7170/iio_lt7170.c:
```diff
@@ -0,0 +1,1039 @@
+	END_ATTRIBUTES_ARRAY
+};
+
+
```
nit: for consistency, just remove this extra new line after the function.

kister-jimenez commented on drivers/power/lt7170/README.rst:
```diff
@@ -0,0 +1,287 @@
+LT7170 Driver Initialization Example
+-------------------------------------
+
+.. code-block:: bash
```
Should this be c instead of bash?

csarip replied:
Used Bash for consistency, as other drivers use Bash instead of C.

kister-jimenez commented on drivers/power/lt7170/README.rst:
```diff
@@ -0,0 +1,287 @@
+LT7170 IIO Driver Initialization Example
+-----------------------------------------
+
+.. code-block:: bash
```
ditto

kister-jimenez commented on projects/lt7170/src/common/common_data.c:
```diff
@@ -0,0 +1,71 @@
+	.chip_id = ID_LT7170,
+	.external_clk_en = false,
+	.crc_en = false,
+};
```
Add a newline at the end of the file.

kister-jimenez reviewed (changes requested):
Please use the 50/72 rule for the commit subject/body

CiprianRegus commented on drivers/power/lt7170/lt7170.c:
```diff
@@ -0,0 +1,1074 @@
+		goto i2c_err;
+
+	/* Identify device */
+	ret = lt7170_read_word(dev, LT7170_MFR_SPECIAL_ID, &word);
+	if (word != LT7170_SPECIAL_ID_VALUE) {
+		ret = -EIO;
+		goto dev_err;
+	}
```
Do we need to check this? If the MCU can't communicate with the device, the next check would fail anyway, right?

CiprianRegus commented on drivers/power/lt7170/lt7170.c:
```diff
@@ -0,0 +1,1074 @@
+
+		*data = rx_buf[0];
+		return ret;
+	} else
```
add `{` for the `else` branch as well if the first case required it.

CiprianRegus commented on drivers/power/lt7170/iio_lt7170.c:
```diff
@@ -0,0 +1,1039 @@
+	case LT7170_MFR_SYNC_CONFIG:
+	case LT7170_MFR_RAIL_ADDRESS:
+	case LT7170_MFR_DISABLE_OUTPUT:
+		return lt7170_read_byte(lt7170, (uint8_t)reg,
+					(uint8_t *)readval);
```
You will have to use a separate `uint8_t` variable for this. Currently, the result will be different on big endian architectures

csarip replied:
Hi, I apologize I'm a bit confused. Could you kindly elaborate on this? I referred to this implementation on that line of code:  [LT7182S_IIO](https://github.com/analogdevicesinc/no-OS/blob/main/drivers/power/lt7182s/iio_lt7182s.c#L123)

CiprianRegus replied:
The function call should be done as following:

```
uint8_t reg_val;
lt7170_read_byte(lt7170, reg, &readval);
*read_val = reg_val;
```

Since your `lt7170_read_byte()` function takes a parameter that's a pointer to an 1 byte unsigned int, it will end up setting the first byte at the address `readval`. Depending on the microcontroller endianess, the actual result will be interpreted as `0xFE000000` or `0x000000FE`, so the values will be different.

Another issue is that by setting only 1 byte, you don't clear any non 0 values in that memory region, so the result will be wrong. For example:

uint32_t reg_val; /* Not initialized, so its value is set to whatever was on the stack */
lt7170_iio_reg_read(..., &reg_val);

At this point, only the first byte of `reg_val` was set, so your result will be wrong.

CiprianRegus commented on drivers/power/lt7170/iio_lt7170.c:
```diff
@@ -0,0 +1,1039 @@
+		.show = lt7170_iio_read_vout,
+		.store = lt7170_iio_write_vout,
+		.priv = LT7170_VOUT_COMMAND
+	},
+	{
+		.name = "vout_max",
+		.show = lt7170_iio_read_vout,
+		.store = lt7170_iio_write_vout,
+		.priv = LT7170_VOUT_MAX
+	},
+	{
+		.name = "vout_margin_high",
+		.show = lt7170_iio_read_vout,
+		.store = lt7170_iio_write_vout,
+		.priv = LT7170_VOUT_MARGIN_HIGH
+	},
+	{
+		.name = "vout_margin_low",
+		.show = lt7170_iio_read_vout,
+		.store = lt7170_iio_write_vout,
+		.priv = LT7170_VOUT_MARGIN_LOW
+	},
+	{
+		.name = "phase",
+		.show = lt7170_iio_read_pwm,
+		.store = lt7170_iio_write_pwm,
+		.priv = LT7170_IIO_PWM_PHASE
+	},
+	{
+		.name = "pulse_skipping",
+		.show = lt7170_iio_read_enable,
+		.store = lt7170_iio_write_enable,
+		.priv = LT7170_IIO_PULSE_ENABLE
+	},
+	{
+		.name = "pulse_skipping_available",
+		.show = lt7170_iio_read_enable_available,
+		.shared = IIO_SHARED_BY_ALL
+	},
+	END_ATTRIBUTES_ARRAY
```
Most of these are not IIO attributes defined in the IIO ABI (https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-bus-iio). Thus, they should be debug attributes instead.

csarip commented:
V2:

- Corrected formatting as per review comments.
- Replaced dynamically allocated array with stack-allocated array.
- Removed unnecessary checking of device.
- Used a separate uint8_t variable to ensure correct result regardless of the microcontroller's endianness.
- Moved attributes that belong to debug attributes.
